docs/config.md
A `config/atc.yml` file contains two sections: `server` and `routes`.

### server

This section is completely optional.  Every field listed below has a default, and can be overriddenby setting an environment variable with the same name, just in upper case.

| Field Name            | Description                                             | Default Value |
|-----------------------|---------------------------------------------------------|---------------|
| `target_port`         | The port that your Puma server should run on. Thruster will set `PORT` to this value when starting your server. | 3000 |
| `cache_size`          | The size of the HTTP cache in bytes. | 64MB |
| `max_cache_item_size` | The maximum size of a single item in the HTTP cache in bytes. | 1MB |
| `x_sendfile_enabled`  | Whether to enable X-Sendfile support. Set to `0` or `false` to disable. | Enabled |
| `max_request_body`    | The maximum size of a request body in bytes. Requests larger than this size will be refused; `0` means no maximum size is enforced. | `0` |
| `bad_gateway_page`    | Path to an HTML file to serve when the backend server returns a 502 Bad Gateway error. If there is no file at the specific path, Thruster will serve an empty 502 response instead. | `./public/502.html` |
| `http_port`           | The port to listen on for HTTP traffic. | 3000 (development)<br>8080 (production) |
| `http_idle_timeout`   | The maximum time in seconds that a client can be idle before the connection is closed. | 60 |
| `http_read_timeout`   | The maximum time in seconds that a client can take to send the request headers and body. | 30 |
| `http_write_timeout`  | The maximum time in seconds during which the client must read the response. | 30 |
| `forward_headers`     | Whether to forward X-Forwarded-* headers from the client. | Disabled when running with TLS; enabled otherwise |
| `debug`               | Set to `1` or `true` to enable debug logging. | Disabled |
| `health_check_path`   | Path used to check if the app is ready to accept requests | `/up` |

### routes

An array of routes to tenant applications.  Each contains:

| Field Name            | Description                                             | Default Value |
|-----------------------|---------------------------------------------------------|---------------|
| `name`                | Name of route.  Used to create PID file and passed to the application in the `FLY_ATC_NAME` environment variable. |
| `endpoint`            | Prefix path.  Used to route requests and passed to the application in the `FLY_ATC_SCOPE` environment variable. |
| `database`            | Name of the database.  `.sqlite3` is appended to this name, and the result is used to replace the basename specified in `DATABASE_URL`. |
| `region`              | Region where this application is hosted. |
| `instance`            | Machine id where this application is hosted.  Takes precedence over `region` |

If no routes are specified, a single route with a blank name and endpoint is created.



internal/config.go
package internal

import (
	"errors"
	"log/slog"
	"os"
	"strconv"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

const (
	configFile = "config/atc.yml"

	KB = 1024
	MB = 1024 * KB

	ENV_PREFIX = "THRUSTER_"

	defaultTargetPort = 0

	defaultCacheSize             = 64 * MB
	defaultMaxCacheItemSizeBytes = 1 * MB
	defaultMaxRequestBody        = 0

	defaultBadGatewayPage = "./public/502.html"

	defaultHttpPort         = 3000
	defaultHttpIdleTimeout  = 60 * time.Second
	defaultHttpReadTimeout  = 30 * time.Second
	defaultHttpWriteTimeout = 30 * time.Second

	defaultLogLevel = slog.LevelInfo

	defaultHealthCheckPath = "/up"
)

type Config struct {
	TargetPort      int      `yaml:"target_port"`
	UpstreamCommand string   `yaml:"upstream_command"`
	UpstreamArgs    []string `yaml:"upstream_args"`

	CacheSizeBytes        int  `yaml:"cache_size"`
	MaxCacheItemSizeBytes int  `yaml:"max_cache_item_size"`
	XSendfileEnabled      bool `yaml:"x_sendfile_enabled"`
	MaxRequestBody        int  `yaml:"max_request_body"`

	BadGatewayPage string `yaml:"bad_gateway_page"`

	HttpPort         int           `yaml:"http_port"`
	HttpIdleTimeout  time.Duration `yaml:"http_idle_timeout"`
	HttpReadTimeout  time.Duration `yaml:"http_read_timeout"`
	HttpWriteTimeout time.Duration `yaml:"http_write_timeout"`

	HealthCheckPath string `yaml:"health_check_path"`

	LogLevel slog.Level `yaml:"log_level"`
}

type Route struct {
	Name     string `yaml:"name"`
	Endpoint string `yaml:"endpoint"`

	Database string `yaml:"database"`
	Region   string `yaml:"region"`
	Instance string `yaml:"instance"`

	Monitor *Monitor
}

type Settings struct {
	Server *Config `yaml:"server"`
	Routes []Route `yaml:"routes"`
}

var settings Settings

func Routes() []Route {
	return settings.Routes
}

func NewConfig() (*Config, error) {
	config := &Config{
		TargetPort: defaultTargetPort,

		CacheSizeBytes:        defaultCacheSize,
		MaxCacheItemSizeBytes: defaultMaxCacheItemSizeBytes,
		XSendfileEnabled:      true,
		MaxRequestBody:        defaultMaxRequestBody,

		BadGatewayPage: defaultBadGatewayPage,

		HttpPort:         defaultHttpPort,
		HttpIdleTimeout:  defaultHttpIdleTimeout,
		HttpReadTimeout:  defaultHttpReadTimeout,
		HttpWriteTimeout: defaultHttpWriteTimeout,

		LogLevel: defaultLogLevel,

		HealthCheckPath: defaultHealthCheckPath,
	}

	if os.Getenv("RAILS_ENV") == "production" {
		config.HttpPort = 8080
	}

	settings.Server = config

	data, err := os.ReadFile(configFile)
	if err == nil {
		err = yaml.Unmarshal(data, &settings)
		if err != nil {
			return nil, err
		}
	}

	if getEnvBool("DEBUG", false) {
		config.LogLevel = slog.LevelDebug
	}

	config.TargetPort = getEnvInt("TARGET_PORT", config.TargetPort)
	config.CacheSizeBytes = getEnvInt("CACHE_SIZE", config.CacheSizeBytes)
	config.MaxCacheItemSizeBytes = getEnvInt("MAX_CACHE_ITEM_SIZE", config.MaxCacheItemSizeBytes)
	config.XSendfileEnabled = getEnvBool("X_SENDFILE_ENABLED", config.XSendfileEnabled)
	config.MaxRequestBody = getEnvInt("MAX_REQUEST_BODY", config.MaxRequestBody)
	config.BadGatewayPage = getEnvString("BAD_GATEWAY_PAGE", config.BadGatewayPage)
	config.HttpPort = getEnvInt("HTTP_PORT", config.HttpPort)
	config.HttpIdleTimeout = getEnvDuration("HTTP_IDLE_TIMEOUT", config.HttpIdleTimeout)
	config.HttpReadTimeout = getEnvDuration("HTTP_READ_TIMEOUT", config.HttpReadTimeout)
	config.HttpWriteTimeout = getEnvDuration("HTTP_WRITE_TIMEOUT", config.HttpWriteTimeout)
	config.LogLevel = slog.Level(getEnvInt("LOG_LEVEL", int(config.LogLevel)))
	config.HealthCheckPath = getEnvString("HEALTH_CHECK_PATH", config.HealthCheckPath)

	if len(os.Args) >= 2 {
		config.UpstreamCommand = os.Args[1]
		config.UpstreamArgs = os.Args[2:]
	} else if config.UpstreamCommand == "" {
		return nil, errors.New("missing upstream command")
	}

	if len(settings.Routes) == 0 {
		settings.Routes = append(settings.Routes, Route{
			Name:     "",
			Endpoint: "",
			Database: getEnvString("RAILS_ENV", "development"),
		})
	}

	return config, nil
}

func findEnv(key string) (string, bool) {
	value, ok := os.LookupEnv(ENV_PREFIX + key)
	if ok {
		return value, true
	}

	value, ok = os.LookupEnv(key)
	if ok {
		return value, true
	}

	return "", false
}

func getEnvString(key, defaultValue string) string {
	value, ok := findEnv(key)
	if ok {
		return value
	}

	return defaultValue
}

func getEnvStrings(key string, defaultValue []string) []string {
	value, ok := findEnv(key)
	if ok {
		items := strings.Split(value, ",")
		result := []string{}

		for _, item := range items {
			item = strings.TrimSpace(item)
			if item != "" {
				result = append(result, item)
			}
		}

		return result
	}

	return defaultValue
}

func getEnvInt(key string, defaultValue int) int {
	value, ok := findEnv(key)
	if !ok {
		return defaultValue
	}

	intValue, err := strconv.Atoi(value)
	if err != nil {
		return defaultValue
	}

	return intValue
}

func getEnvDuration(key string, defaultValue time.Duration) time.Duration {
	value, ok := findEnv(key)
	if !ok {
		return defaultValue
	}

	intValue, err := strconv.Atoi(value)
	if err != nil {
		return defaultValue
	}

	return time.Duration(intValue) * time.Second
}

func getEnvBool(key string, defaultValue bool) bool {
	value, ok := findEnv(key)
	if !ok {
		return defaultValue
	}

	boolValue, err := strconv.ParseBool(value)
	if err != nil {
		return defaultValue
	}

	return boolValue
}
